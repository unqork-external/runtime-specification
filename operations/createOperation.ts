import type { OperationTypes } from './enums/operation-types.enum'
import { OperationTypeMap } from './typemaps/operation.typemap'
import type { CustomOperationTypeMap } from '../../operations'
import type { InputRef } from '../inputs/inputRef'

/**
 * Given an OperationType, Typescript will complete the options list.
 *
 * In return, the caller receives Typescript constructed class.
 *
 * This includes preset defaults.
 *
 * In essence, this function ensures that all keys are fully present, fully typed, and ready for runtime
 * validation if necessary.
 *
 * @param type
 * @param options
 * @param name name of the operation
 * @param additional
 * @param spreadOp
 * @param additionalArgs
 * @returns - The Operation Class type
 */
export const createOperation = <OpType extends OperationTypes>(
  type: OpType,
  options: InstanceType<(typeof OperationTypeMap)[OpType]>['options'],
  additionalArgs?: Partial<AdditionalArgs>,
): InstanceType<(typeof OperationTypeMap)[OpType]> => {
  const classOpType = OperationTypeMap[type]
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

export const createCustomOperation = <OperationEnum extends string, OpType extends OperationEnum>(
  type: OpType,
  options: InstanceType<CustomOperationTypeMap<OperationEnum>[OpType]['model']>['options'],
  typeMap: CustomOperationTypeMap<OperationEnum>,
  additionalArgs?: Partial<AdditionalArgs>,
) => {
  const classOpType = typeMap[type].model
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

interface AdditionalArgs {
  name: string
  notifyImmediately: boolean
  creatorSummary: string
}
