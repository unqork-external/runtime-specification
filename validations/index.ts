export * from './validation.enum'
export * from './handlerTypes'
export * from './validation'
export * from './validationResult'
export * from './validationRule.type'
export * from './validationError.type'
export * from './ruleToValue.typemap'
