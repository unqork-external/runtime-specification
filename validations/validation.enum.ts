// NOTE: Keeping values camel case to make working with utils imported from R1 easier
export enum ValidationRuleTypes {
  REQUIRED = 'required',
  MAX_LENGTH = 'maxLength',
  MIN_LENGTH = 'minLength',
  ARRAY_MIN_LENGTH = 'arrayMinLength',
  ARRAY_MAX_LENGTH = 'arrayMaxLength',
  PATTERN = 'pattern',
  MIN = 'min',
  MAX = 'max',
  MASK = 'mask',
  NUMBER = 'number',
  // dateinput is intentionally all lower case
  DATE_INPUT = 'dateinput',
  IS_BEFORE_DATE = 'isBeforeDate',
  IS_AFTER_DATE = 'isAfterDate',
  CUSTOM_ERROR = 'customError',
  IS_VALID_EMAIL = 'isEmail',
}
