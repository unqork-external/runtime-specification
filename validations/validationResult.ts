import { Nullable, CollectionOf, Property } from '@tsed/schema'

import { ValidationError } from './validationError.type'

export class ValidationResult {
  //TODO: Change null to undefined
  @Nullable(Boolean)
  isValid?: boolean | null = null

  @CollectionOf(ValidationError)
  errors?: ValidationError[]
}

export class ModuleValidationResult {
  @Property()
  isValid: boolean

  @CollectionOf(ValidationResult)
  results: Record<string, ValidationResult>
}
