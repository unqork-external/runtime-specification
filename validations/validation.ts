import { Description, Example, Property } from '@tsed/schema'

import { ValidationResult } from './validationResult'
import type { ValidationRules } from './validationRule.type'
import { trimAll } from '../../utilities'

export class Validation {
  @Property(ValidationResult)
  @Description(`Describes the current state of validation for this component`)
  result?: ValidationResult

  @Property()
  @Description(
    trimAll(`
      The set of rules to restrict or define the accepted data for the current component.
      These are the rules to run for validation of the data input.
    `),
  )
  @Example(
    trimAll(`
      You can use it to define the max length of characters in an input field with a custom error message.
      rules: {
        required: {
          type: ValidationRuleTypes.MAX_LENGTH,
          validation: {
            maxLength: 5,
          },
          errorMessage: 'Max length is 5',
        },
      },
    `),
  )
  rules?: ValidationRules
}
