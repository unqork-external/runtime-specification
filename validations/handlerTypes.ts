import { ValidationRuleTypes } from './validation.enum'

export type ValidationHandlerValidationsMap = {
  [ValidationRuleTypes.REQUIRED]: { required: boolean }
  [ValidationRuleTypes.PATTERN]: { pattern: RegExp }
  [ValidationRuleTypes.MAX_LENGTH]: { maxLength: number }
  [ValidationRuleTypes.MIN_LENGTH]: { minLength: number }
  [ValidationRuleTypes.ARRAY_MAX_LENGTH]: { arrayMaxLength: number }
  [ValidationRuleTypes.ARRAY_MIN_LENGTH]: { arrayMinLength: number }
  [ValidationRuleTypes.MIN]: { min: string }
  [ValidationRuleTypes.MAX]: { max: string }
  [ValidationRuleTypes.NUMBER]: { number: boolean }
  [ValidationRuleTypes.IS_BEFORE_DATE]: { targetDate: string }
  [ValidationRuleTypes.IS_AFTER_DATE]: { targetDate: string }
  [ValidationRuleTypes.DATE_INPUT]: { dateinput: string }
  [ValidationRuleTypes.CUSTOM_ERROR]: { customError: boolean }
  [ValidationRuleTypes.IS_VALID_EMAIL]: { isEmail: boolean }
  [ValidationRuleTypes.MASK]: { mask: string }
}
/**
 * TODO: What are the types??
 *
 * Should return true when its valid and false when its invalid
 */
export type ValidationHandler = (value: any, validation: any) => boolean

export type ValidationHandlers = Record<ValidationRuleTypes, ValidationHandler>
