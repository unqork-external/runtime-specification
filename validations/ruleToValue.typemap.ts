import { ValidationRuleTypes } from './validation.enum'

export type ValidationRuleValidationsValues = {
  [ValidationRuleTypes.REQUIRED]: boolean
  [ValidationRuleTypes.MAX_LENGTH]: number
  [ValidationRuleTypes.MIN_LENGTH]: number
  [ValidationRuleTypes.ARRAY_MAX_LENGTH]: unknown[]
  [ValidationRuleTypes.ARRAY_MIN_LENGTH]: unknown[]
  [ValidationRuleTypes.PATTERN]: RegExp
  [ValidationRuleTypes.MIN]: string
  [ValidationRuleTypes.MAX]: string
  [ValidationRuleTypes.NUMBER]: boolean
  // Note this could change base on how the date input component is crafted. For now leaving as a string
  [ValidationRuleTypes.DATE_INPUT]: string
  [ValidationRuleTypes.IS_BEFORE_DATE]: string
  [ValidationRuleTypes.IS_AFTER_DATE]: string
  [ValidationRuleTypes.IS_VALID_EMAIL]: boolean
}
