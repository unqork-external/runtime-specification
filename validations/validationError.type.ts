import { Description, Enum, Required, Optional } from '@tsed/schema'

import { ValidationRuleTypes } from './validation.enum'

export class ValidationError {
  @Required()
  @Enum(ValidationRuleTypes)
  @Description('Type of the error')
  type: ValidationRuleTypes

  @Optional()
  @Description('The error message')
  message?: string
}
