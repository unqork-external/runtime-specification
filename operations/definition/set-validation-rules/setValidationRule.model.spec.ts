import { SetValidationRuleOperation } from './setValidationRule.model'
import { generateSchemaAndValidate } from '../../../../utilities'
import { ValidationRuleTypes } from '../../../validations'
import { createOperation } from '../../createOperation'
import { OperationTypes } from '../../enums/operation-types.enum'

describe('setValidation', () => {
  let validate, schema

  beforeAll(() => {
    const response = generateSchemaAndValidate(SetValidationRuleOperation)
    validate = response.validate
    schema = response.schema
  })

  it('should match JSON schema', () => {
    expect(schema).toMatchSnapshot()
  })

  it('should validate submit operation with all required keys', () => {
    const op = createOperation(OperationTypes.SET_VALIDATION_RULE, {
      targetKey: '1',
      rule: {
        type: ValidationRuleTypes.DATE_INPUT,
      },
    })
    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })

  it('should return false if validation rules are not being send', () => {
    //@ts-ignore
    const op = createOperation(OperationTypes.SET_VALIDATION_RULE, {})
    const isValid = validate(op)
    expect(isValid).toBeFalse()
  })
})
