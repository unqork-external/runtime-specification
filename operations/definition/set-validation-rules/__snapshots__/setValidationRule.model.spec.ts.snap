// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setValidation should match JSON schema 1`] = `
{
  "definitions": {
    "SetValidationRuleOptions": {
      "properties": {
        "rule": {
          "$ref": "#/definitions/ValidationRule",
        },
        "shouldExecute": {
          "default": true,
          "description": "
    Conditionally execute this operation.
    Can set to false to temporarily disabled when debugging.
  ",
          "type": [
            "string",
            "boolean",
          ],
        },
        "targetKey": {
          "description": "Key of target to apply set validation rule",
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "targetKey",
      ],
      "type": "object",
    },
    "ValidationRule": {
      "properties": {
        "errorMessage": {
          "description": "What message to show when this rule is not met",
          "type": "string",
        },
        "type": {
          "description": "What type of validation rule is this",
          "enum": [
            "required",
            "maxLength",
            "minLength",
            "arrayMinLength",
            "arrayMaxLength",
            "pattern",
            "min",
            "max",
            "mask",
            "number",
            "dateinput",
            "isBeforeDate",
            "isAfterDate",
            "customError",
            "isEmail",
          ],
          "type": "string",
        },
        "validation": {
          "description": "Nested validation record",
          "type": "object",
        },
      },
      "type": "object",
    },
  },
  "properties": {
    "creatorSummary": {
      "description": "A detailed summary of the operation",
      "type": "string",
    },
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/SetValidationRuleOptions",
    },
    "type": {
      "const": "SET_VALIDATION_RULE",
      "examples": [
        "SET_VALIDATION_RULE",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "type": "object",
}
`;
